### PR TITLE
Use cached chunk size in ExpressionEvaluator to fix multithreaded bottleneck

### PR DESCRIPTION
**Pull request by lanice**
Some TPC-H queries show a significant drop in performance when scaling up the number of used cores, such as TPC-H 7 and 19. In the ExpressionEvaluator, we use `_chunk->size()` as maximum of some for-loops. `Chunk::size()` in turn calls `Chunk::get_segment()`, which has an `std::atomic_load()`. That causes a major bottleneck the more threads simultaneously try to do that. Just replacing `_chunk->size()` with the `_output_row_count` that we have anyway gives us a small performance boost:
(nemea: `./build-release/hyriseBenchmarkTPCH --scale 1 --scheduler --clients 10 --cores 224`)
```
+-----------+-------------------+-------+--------------------+-------+---------+-----------------------------------------------+
| Benchmark | prev. iter/s      | runs  | new iter/s         | runs  | change  |               p-value (significant if <0.001) |
+-----------+-------------------+-------+--------------------+-------+---------+-----------------------------------------------+
| TPC-H 1   | 4.36615133285522  | 262   | 4.3499813079834    | 261   | -0%     |                                        0.9405 |
| TPC-H 2   | 1.59978568553925  | 96    | 1.76655912399292   | 106   | +10%    |                                        0.2186 |
| TPC-H 3   | 52.2665786743164  | 3136  | 52.376392364502    | 3143  | +0%     |                                        0.5229 |
| TPC-H 4   | 33.4331398010254  | 2006  | 33.3294792175293   | 2000  | -0%     |                                        0.4619 |
| TPC-H 5   | 30.7287921905518  | 1844  | 30.9787864685059   | 1859  | +1%     |                                        0.1492 |
| TPC-H 6   | 196.972305297852  | 10001 | 196.706588745117   | 10000 | -0%     |                                        0.4295 |
| TPC-H 7   | 0.799872219562531 | 48    | 17.6814384460449   | 1061  | +2111%  |                                        0.0000 |
| TPC-H 8   | 44.7800216674805  | 2687  | 48.5134963989258   | 2911  | +8%     |                                        0.0000 |
| TPC-H 9   | 17.6331195831299  | 1058  | 17.7157974243164   | 1063  | +0%     |                                        0.4249 |
| TPC-H 10  | 18.1639080047607  | 1090  | 18.166524887085    | 1090  | +0%     |                                        0.9713 |
| TPC-H 11  | 118.83235168457   | 7130  | 122.702835083008   | 7363  | +3%     |                                        0.0000 |
| TPC-H 12  | 37.9322471618652  | 2276  | 96.2268753051758   | 5774  | +154%   |                                        0.0000 |
| TPC-H 13  | 14.7490749359131  | 885   | 14.4994840621948   | 870   | -2%     |                                        0.0101 |
| TPC-H 14  | 109.266319274902  | 6557  | 107.996383666992   | 6480  | -1%     |                                        0.0007 |
| TPC-H 16  | 3.1497061252594   | 189   | 44.4944648742676   | 2670  | +1313%  |                                        0.0000 |
| TPC-H 17  | 0.0               | 0     | 0.0                | 0     | nan%    |    (run time too short) (not enough runs) nan |
| TPC-H 18  | 2.99956774711609  | 180   | 10.1654663085938   | 610   | +239%   |                                        0.0000 |
| TPC-H 19  | 0.916580498218536 | 55    | 131.436614990234   | 7887  | +14240% |                                        0.0000 |
| TPC-H 20  | 0.149998962879181 | 9     | 0.0333289206027985 | 2     | -78%    | (run time too short) (not enough runs) 0.0474 |
| TPC-H 21  | 0.0               | 0     | 0.0                | 0     | nan%    |    (run time too short) (not enough runs) nan |
| TPC-H 22  | 20.3646640777588  | 1222  | 72.2428131103516   | 4335  | +255%   |                                        0.0000 |
| average   |                   |       |                    |       | +869%   |                                               |
+-----------+-------------------+-------+--------------------+-------+---------+-----------------------------------------------+
```
Another thing that might be worth looking at is caching the chunksize in the Chunk itself once it becomes immutable. That way we can omit the calls to get_segment once the Chunk is "complete". But that is not necessary for this PR, since `_output_row_count` should be used instead of calling `_chunk->size()` anyway.

### CODE REVIEW

**Comment by mrks** [comment]
How did you find this? Did it show up in the profiler?

File: src/benchmark/tpch_benchmark.cpp [hunk 1]
```diff
@@ -110,7 +110,8 @@ int main(int argc, char* argv[]) {
 
   // TODO(leander): Enable support for queries that contain multiple statements requiring execution
   if (config->enable_scheduler) {
-    Assert(std::find(query_ids.begin(), query_ids.end(), opossum::QueryID{15}) == query_ids.end(),
+    // QueryID{14} represents TPC-H query 15 because of the way we generate and store the queries
```
Comment (mrzzzrm) [inline src/benchmark/tpch_benchmark.cpp]: That sounds cryptic. "QueryID{14} represents TPC-H query 15 because we use 0 indexing", I suppose?

**Comment by lanice** [comment]
Yeah I just did some more VTune experiments on TPC-H 19, to see if it finds anything.. it did :)

File: src/benchmark/tpch_benchmark.cpp (hunk 1 above)
Comment (lanice) [reply src/benchmark/tpch_benchmark.cpp]: Right, that makes sense!

**Comment by Bouncner** [comment]
> gives us a small performance boost:

+-----------+-------------------+-------+--------------------+-------+---------+-----------------------------------------------+
| Benchmark | prev. iter/s      | runs  | new iter/s         | runs  | change  |               p-value (significant if <0.001) |
+-----------+-------------------+-------+--------------------+-------+---------+-----------------------------------------------+
| TPC-H 7   | 0.799872219562531 | 48    | 17.6814384460449   | 1061  | +2111%  |                                        0.0000 |
| TPC-H 16  | 3.1497061252594   | 189   | 44.4944648742676   | 2670  | +1313%  |                                        0.0000 |
| TPC-H 18  | 2.99956774711609  | 180   | 10.1654663085938   | 610   | +239%   |                                        0.0000 |
| TPC-H 19  | 0.916580498218536 | 55    | 131.436614990234   | 7887  | +14240% |                                        0.0000 |
| TPC-H 21  | 0.0               | 0     | 0.0                | 0     | nan%    |    (run time too short) (not enough runs) nan |
| TPC-H 22  | 20.3646640777588  | 1222  | 72.2428131103516   | 4335  | +255%   |                                        0.0000 |
+-----------+-------------------+-------+--------------------+-------+---------+-----------------------------------------------+

I like your humor.

**Comment by lanice** [comment]
I think the problem is that the compiler can't necessarily optimize the end condition function call away, if it's not a built-in function. Which makes sense, because the compiler can't assume that the output of _chunk->size() doesn't change between iterations. So it has to be called for each iteration of the loop.

**Comment by Bouncner** [comment]
Jep, makes totally sense (we just added this two the developer hints two weeks ago). I was confused this morning. 😇

Still, nice improvements!

**Comment by mrks** [comment]
FYI - this even helped the single-threaded execution (sf 0,1):

```
 Benchmark | prev. iter/s | new iter/s | change
-----------+--------------+------------+----------------
 TPC-H 1   | 6.245        | 6.21       | -1%
 TPC-H 2   | 130.368      | 133.346    | +2%
 TPC-H 3   | 66.315       | 62.216     | -6%
 TPC-H 4   | 23.904       | 23.581     | -1%
 TPC-H 5   | 50.596       | 50.192     | -1%
 TPC-H 6   | 187.477      | 192.292    | +3%
 TPC-H 7   | 13.509       | 21.315     | +58%
 TPC-H 8   | 85.54        | 86.341     | +1%
 TPC-H 9   | 26.351       | 26.306     | -0%
 TPC-H 10  | 34.397       | 33.873     | -2%
 TPC-H 11  | 206.791      | 220.087    | +6%
 TPC-H 12  | 61.035       | 63.781     | +4%
 TPC-H 13  | 28.313       | 28.587     | +1%
 TPC-H 14  | 186.812      | 190.899    | +2%
 TPC-H 15  | 86.385       | 90.281     | +5%
 TPC-H 16  | 55.026       | 69.887     | +27%
 TPC-H 17  | 7.305        | 7.15       | -2%
 TPC-H 18  | 17.355       | 19.353     | +12%
 TPC-H 19  | 25.624       | 59.471     | +132%
 TPC-H 20  | 3.977        | 4.112      | +3%
 TPC-H 21  | 1.972        | 2.082      | +6%
 TPC-H 22  | 106.374      | 124.38     | +17%
 average   |              |            | +12%
```

**Comment by lanice** [comment]
Oh interesting, nice!

**Pull request merged**